keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,704 @@
+    participant node
+    participant kubelet
+    participant cAdvisor-cache
+    participant machine-info
+    kubelet->>cAdvisor-cache: Poll
+    cAdvisor-cache->>machine-info: fetch
+    machine-info->>cAdvisor-cache: update
+    cAdvisor-cache->>kubelet: update
+    kubelet->>node: node status update
+    kubelet->>node: re-initialize resource managers
+```
+
+The interaction sequence is as follows
+1. Kubelet will be polling in interval to fetch the machine resource information from cAdvisor's cache, Which is currently updated every 5 minutes.
+2. Kubelet's cache will be updated with the latest machine resource information.
+3. Node status updater will update the node's status with the latest resource information.
+4. Kubelet will reinitialize the resource managers to keep them up to date with dynamic resource changes.
+
+With increase in cluster resources the following components will updated
+
+1. Scheduler
+   * Scheduler will automatically schedule any pending pods.
+
+
+2. Change in Swap Memory limit
+   * Currently, the swap memory limit is calculated as 
+ `(<containerMemoryRequest>/<nodeTotalMemory>)*<totalPodsSwapAvailable>`
+   * So increase in nodeTotalMemory will result in updated swap memory limit.
+
+
+3. Change in OOM score
+   * OOM score calculation depends on machine's memory, so the new OOM score will be updated accordingly.
+
+**Proposed Code changes**
+
+**Dynamic Node Scale Up logic**
+
+```go
+	if utilfeature.DefaultFeatureGate.Enabled(features.NodeResourceHotPlug) {
+		// Handle the node dynamic scale up
+		machineInfo, err := kl.cadvisor.MachineInfo()
+		if err != nil {
+			klog.ErrorS(err, "Error fetching machine info")
+		} else {
+			cachedMachineInfo, _ := kl.GetCachedMachineInfo()
+            // Avoid collector collects it as a timestamped metric
+            // See PR #95210 and #97006 for more details.
+                machineInfo.Timestamp = time.Time{}
+			if !reflect.DeepEqual(cachedMachineInfo, machineInfo) {
+				kl.setCachedMachineInfo(machineInfo)
+
+				// Resync the resource managers
+				if err := kl.ResyncComponents(machineInfo); err != nil {
+					klog.ErrorS(err, "Error resyncing the kubelet components with machine info")
+				}
+			}
+		}
+	}
+```
+
+**Changes to resource managers to adapt to dynamic scale up of resources**
+
+1. Adding ResyncComponents() method to ContainerManager interface
+```go
+    // Manages the containers running on a machine.
+    type ContainerManager interface {
+        .
+        .
+        // ResyncComponents will resyc the resource managers like cpu, memory and topology managers
+	// with updated machineInfo
+	ResyncComponents(machineInfo *cadvisorapi.MachineInfo) error
+	.
+	.
+    )
+```
+
+2. Adding a method Sync to all the resource managers and will be invoked once there is dynamic resource change.
+
+```go
+        // SyncMachineInfo will sync the Manager with the latest machine info
+        SyncMachineInfo(machineInfo *cadvisorapi.MachineInfo) error
```

Comment:
we need to discuss if this should or should not cause a workload rebalance. I think no, but this should be elaborated in the KEP.